sdk/src/test/java/com/qonversion/android/sdk/internal/userProperties/UserPropertiesStorageImplTest.kt

```diff
@@ -0,0 +1,451 @@
+
+            // when
+            val result = userPropertiesStorage.getPropertiesFromStorage()
+
+            //then
+            assertThat(result).isEqualTo(result)
+        }
+
+        @Test
+        fun `get properties from empty storage`() {
+            // given
+            every {
+                mockLocalStorage.getString(propertiesKeyInMemory)
+            } returns null
+
+            // when
+            val result = userPropertiesStorage.getPropertiesFromStorage()
+
+            //then
+            verify { mockMapper wasNot called }
+            assertThat(result).isEmpty()
+        }
+
+        @Test
+        fun `get properties when exception occurred`() {
+            // given
+            val errorString = "Couldn't load properties from storage"
+            val jsonString = "jsonString"
+            every {
+                mockLocalStorage.getString(propertiesKeyInMemory)
+            } returns jsonString
+
+            val exception = IllegalStateException("Couldn't create JSONObject from string")
+            every {
+                mockMapper.toMap(jsonString)
+            } throws exception
+
+            // when
+            userPropertiesStorage.getPropertiesFromStorage()
+            assertThat(slotErrorLogMessage.captured)
+                .startsWith(errorString)
+            verify(exactly = 1) {
+                mockLogger.error(errorString, exception)
+            }
+        }
+    }
+
+    @Nested
+    inner class PutPropertiesToStorage {
+        @Test
+        fun `put properties from filled storage`() {
+            // given
+            val jsonString = "jsonString"
+            val existingProperties = mapOf("key1" to "value1", "key2" to "value2")
+
+            mockGetPropertiesFromStorage(existingProperties)
+
+            every {
+                mockMapper.fromMap(existingProperties)
+            } returns jsonString
+
+            every {
+                mockLocalStorage.putString(propertiesKeyInMemory, jsonString)
+            } just runs
+
+            // when
+            userPropertiesStorage.putPropertiesToStorage()
+
+            //then
+            verify(exactly = 1) {
+                mockLocalStorage.putString(propertiesKeyInMemory, jsonString)
+            }
+        }
+
+        @Test
+        fun `put properties when exception occurred`() {
+            // given
+            val errorString = "Couldn't save properties to storage"
+            val existingProperties = mapOf("key1" to "value1", "key2" to "value2")
+            mockGetPropertiesFromStorage(existingProperties)
+
+            val exception = IllegalStateException("Couldn't create JSONObject from map")
+            every {
+                mockMapper.fromMap(existingProperties)
+            } throws exception
+
+            // when
+            userPropertiesStorage.putPropertiesToStorage()
+
+            //then
+            verify(exactly = 0) {
+                mockLocalStorage.putString(any(), any())
+            }
```

Comment:
Can be replaced if we mock `properties` instead of `getPropertiesFromStorage`
```suggestion
            verify {
                mockLocalStorage wasNot called
            }
```